OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitYoutubeVideoFragment.java

```diff
@@ -0,0 +1,195 @@
+        if (orientation == Configuration.ORIENTATION_LANDSCAPE && youTubePlayer != null) {
+            youTubePlayer.setFullscreen(true);
+        } else if (youTubePlayer != null) {
+            youTubePlayer.setFullscreen(false);
+        }
+    }
+
+    @Override
+    public void onPageShow() {
+        super.onPageShow();
+        initializeYoutubePlayer();
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        subtitleDisplayHandler.removeCallbacks(subtitlesProcessorRunnable);
+    }
+
+    @Override
+    public void onPageDisappear() {
+        super.onPageDisappear();
+        try {
+            getChildFragmentManager().beginTransaction().replace(R.id.player_container, new Fragment()).commit();
+        } catch (IllegalStateException ex) {
+            logger.error(ex);
+        }
+        subtitleDisplayHandler.removeCallbacks(subtitlesProcessorRunnable);
+    }
+
+    @Override
+    public void onInitializationSuccess(Provider provider,
+                                        YouTubePlayer player,
+                                        boolean wasRestored) {
+        final int orientation = getResources().getConfiguration().orientation;
+        if (!wasRestored) {
+            Uri uri = Uri.parse(unit.getData().encodedVideos.getYoutubeVideoInfo().url);
+            String v = uri.getQueryParameter("v");
+            player.loadVideo(v);
+            youTubePlayer = player;
+            if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
+                youTubePlayer.setFullscreen(true);
+            }
+        }
+        subtitleDisplayHandler = new Handler();
+        subtitleDisplayHandler.post(subtitlesProcessorRunnable);
+
+    }
+
+    @Override
+    public void onInitializationFailure(YouTubePlayer.Provider provider,
+                                        YouTubeInitializationResult result) {
+    }
+
+    private Runnable subtitlesProcessorRunnable = new Runnable() {
+        @Override
+        public void run() {
+            if (youTubePlayer != null) {
+                int currentPos = youTubePlayer.getCurrentTimeMillis();
+                if (subtitlesObj != null) {
+                    Collection<Caption> subtitles = subtitlesObj.captions.values();
+                    int currentSubtitleIndex = 0;
+                    for (Caption subtitle : subtitles) {
+                        int startMillis = subtitle.start.getMseconds();
+                        int endMillis = subtitle.end.getMseconds();
+                        if (currentPos >= startMillis && currentPos <= endMillis) {
+                            updateSelection(currentSubtitleIndex);
+                            break;
+                        }
+                        currentSubtitleIndex++;
+                    }
+                }
+            }
+            subtitleDisplayHandler.postDelayed(this, SUBTITLES_DISPLAY_DELAY_MS);
+        }
+    };
+
+    private void setSubtitlesObj(VideoBlockModel unit) {
```

Comment:
I find this function really hard to understand. Is there a way to make it more clear?

Reply:
I tried to reuse the most of code but in this case I couldn't do it, I had to copy the way how the PlayerFragment update the subtitles and use it here, so, I think there is not a way to make that